File: Sorting/10825_국영수.py
#입출력
from sys import stdin
#입력 
input = stdin.readline

t=int(input())
scores=[]

for i in range(t):
    student,kor,eng,mat=input().split()
    scores.append([student,int(kor),int(eng),int(mat)])

#정렬
scores.sort(key = lambda x: (-x[1],x[2],-x[3],x[0]))

#출력
for s in scores:
    print(s[0])

File: Sorting/10825_국영수.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

struct Student {
	string name;
	int kor, eng, mat;

	bool operator < (const Student& s) const {
		if (kor == s.kor) {
			if (eng == s.eng) {
				if (mat == s.mat)
					return name < s.name;
				return mat > s.mat;
			}
			return eng < s.eng;
		}
		return kor > s.kor;
	}
};

void solution(int n, Student* arr) {
	sort(arr, arr + n); // Á¤·Ä

	for (int i = 0; i < n; ++i)
		cout << arr[i].name << '\n';
	return;
}

int main(void) {
	int n;
	Student* arr;

	cin >> n;
	arr = new Student[n];
	for (int i = 0; i < n; ++i) 
		cin >> arr[i].name >> arr[i].kor >> arr[i].eng >> arr[i].mat;
	
	solution(n, arr);
	delete[] arr;
	return 0;
}